cpp/src/copying/gather.cu

```diff
@@ -293,24 +199,59 @@ struct column_gatherer {
   template <typename ColumnType>
   void operator()(gdf_column const* source_column,
                   gdf_index_type const gather_map[],
-                  gdf_column* destination_column, bool check_bounds = false,
-                  cudaStream_t stream = 0) {
+                  gdf_column* destination_column, bool check_bounds,
+                  cudaStream_t stream, bool merge_nvstring_category = false) {
     ColumnType const* const source_data{
         static_cast<ColumnType const*>(source_column->data)};
     ColumnType* destination_data{
         static_cast<ColumnType*>(destination_column->data)};
 
     gdf_size_type const num_destination_rows{destination_column->size};
 
-    // If gathering in-place, allocate temporary buffers to hold intermediate
-    // results
-    bool const in_place{source_data == destination_data};
-    rmm::device_vector<ColumnType> temp_destination;
-    if (in_place) {
-      temp_destination.resize(num_destination_rows);
-      destination_data = temp_destination.data().get();
+    // If gathering in-place or scattering nvstring 
+    // (in which case the merge_nvstring_category should be set to true) allocate 
+    // temporary buffers to hold intermediate results
+    bool const merge_category = std::is_same<ColumnType, nvstring_category>::value && merge_nvstring_category;
+    bool const in_place = !merge_category && (source_data == destination_data);
+    
+    if(merge_category){
+      // merge the categories.
+      gdf_column temp_src = cudf::copy(*source_column);
```

Comment:
This looks terribly inefficient. Why are we making 2 deep copies of both the source and the destination?

Reply:
This is how `sync_column_categories` works ... I was looking at the usage here https://github.com/rapidsai/cudf/blob/9c494ed63d2670d45ae4823e0522e7c51bcb7dcb/cpp/src/copying/copy_range.cu#L93.

Reply:
That example only makes a single copy of the src/dest. Why do you need 2?